expression/builtin_vectorized.go

```diff
@@ -14,13 +14,92 @@
 package expression
 
 import (
+	"sync"
+
 	"github.com/pingcap/errors"
+	"github.com/pingcap/parser/mysql"
 	"github.com/pingcap/tidb/sessionctx"
 	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/types/json"
 	"github.com/pingcap/tidb/util/chunk"
 )
 
+// columnBufferAllocator is used to allocate and release column buffer in vectorized evaluation.
+type columnBufferAllocator interface {
+	// get allocates a column buffer with the specific eval type and capacity.
+	// the allocator is not responsible for initializing the column, so please initialize it before using.
+	get(evalType types.EvalType, capacity int) (*chunk.Column, error)
+	// put releases a column buffer.
+	put(buf *chunk.Column)
+}
+
+type localSliceBuffer struct {
+	sync.Mutex
+	buffers []*chunk.Column
+	head    int
+	tail    int
+	size    int
+}
+
+func newLocalSliceBuffer() *localSliceBuffer {
+	// in most cases, expressions are called serially, so 1 buffer is enough.
+	return &localSliceBuffer{buffers: make([]*chunk.Column, 1)}
```

Comment:
1 seems to be not enough if a function has multiple args.

Reply:
You are right.
Now I add a new argument `initCap` for this function, which is decided by the number of arguments.